File: server/src/modules/chat/chat.repository.ts
import { promises as fs } from "node:fs";
import path from "node:path";
import type { Conversation } from "./chat.types.js";

const CHATS_DIR = path.join(process.cwd(), "data", "chats");

async function ensureChatsDir(): Promise<void> {
  await fs.mkdir(CHATS_DIR, { recursive: true });
}

function conversationFilePath(id: string): string {
  return path.join(CHATS_DIR, `${id}.json`);
}

export class ChatRepository {
  async save(conversation: Conversation): Promise<void> {
    await ensureChatsDir();
    await fs.writeFile(
      conversationFilePath(conversation.id),
      JSON.stringify(conversation, null, 2),
      "utf8"
    );
  }

  async findById(id: string): Promise<Conversation | null> {
    try {
      const raw = await fs.readFile(conversationFilePath(id), "utf8");
      return JSON.parse(raw) as Conversation;
    } catch {
      return null;
    }
  }

  async delete(id: string): Promise<boolean> {
    try {
      await fs.unlink(conversationFilePath(id));
      return true;
    } catch {
      return false;
    }
  }

  async listIds(): Promise<string[]> {
    await ensureChatsDir();
    const entries = await fs.readdir(CHATS_DIR);
    return entries
      .filter((name) => name.endsWith(".json"))
      .map((name) => name.replace(/\.json$/u, ""));
  }

  async listAll(): Promise<Conversation[]> {
    const ids = await this.listIds();
    const conversations: Conversation[] = [];
    for (const id of ids) {
      const conversation = await this.findById(id);
      if (conversation) conversations.push(conversation);
    }
    return conversations.sort(
      (a, b) => new Date(b.updatedAt).getTime() - new Date(a.updatedAt).getTime()
    );
  }
}

File: src-tauri/src/repository/chat_repository.rs
use std::path::{Path, PathBuf};

use tokio::fs;

use crate::models::Conversation;

pub struct ChatRepository;

impl ChatRepository {
    fn file_path(chats_dir: &Path, id: &str) -> PathBuf {
        chats_dir.join(format!("{id}.json"))
    }

    async fn ensure_dir(chats_dir: &Path) -> Result<(), std::io::Error> {
        fs::create_dir_all(chats_dir).await
    }

    pub async fn save(chats_dir: &Path, conversation: &Conversation) -> Result<(), std::io::Error> {
        Self::ensure_dir(chats_dir).await?;
        let json = serde_json::to_string_pretty(conversation).map_err(|e| {
            std::io::Error::new(std::io::ErrorKind::InvalidData, e.to_string())
        })?;
        fs::write(Self::file_path(chats_dir, &conversation.id), json).await
    }

    pub async fn find_by_id(chats_dir: &Path, id: &str) -> Result<Option<Conversation>, std::io::Error> {
        match fs::read_to_string(Self::file_path(chats_dir, id)).await {
            Ok(raw) => {
                let conversation: Conversation = serde_json::from_str(&raw).map_err(|e| {
                    std::io::Error::new(std::io::ErrorKind::InvalidData, e.to_string())
                })?;
                Ok(Some(conversation))
            }
            Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(None),
            Err(e) => Err(e),
        }
    }

    pub async fn delete(chats_dir: &Path, id: &str) -> Result<bool, std::io::Error> {
        match fs::remove_file(Self::file_path(chats_dir, id)).await {
            Ok(()) => Ok(true),
            Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(false),
            Err(e) => Err(e),
        }
    }

    pub async fn list_ids(chats_dir: &Path) -> Result<Vec<String>, std::io::Error> {
        Self::ensure_dir(chats_dir).await?;
        let mut ids = Vec::new();
        let mut entries = fs::read_dir(chats_dir).await?;
        while let Some(entry) = entries.next_entry().await? {
            let name = entry.file_name().to_string_lossy().into_owned();
            if name.ends_with(".json") {
                ids.push(name.trim_end_matches(".json").to_string());
            }
        }
        Ok(ids)
    }

    pub async fn list_all(chats_dir: &Path) -> Result<Vec<Conversation>, std::io::Error> {
        let ids = Self::list_ids(chats_dir).await?;
        let mut conversations = Vec::new();
        for id in ids {
            if let Some(conversation) = Self::find_by_id(chats_dir, &id).await? {
                conversations.push(conversation);
            }
        }
        conversations.sort_by(|a, b| b.updated_at.cmp(&a.updated_at));
        Ok(conversations)
    }
}
